File: library/src/main/java/com/getstream/sdk/chat/logger/StreamChatLogger.java
package com.getstream.sdk.chat.logger;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class StreamChatLogger implements StreamLogger {

    private StreamLoggerLevel loggingLevel;
    private StreamLoggerHandler loggingHandler;

    private StreamChatLogger(@Nullable StreamLoggerLevel loggingLevel, @Nullable StreamLoggerHandler loggingHandler) {
        this.loggingLevel = loggingLevel == null ? StreamLoggerLevel.NOTHING : loggingLevel;
        this.loggingHandler = loggingHandler;
    }

    public void logI(@NonNull Object tag, @NonNull String message) {
        if (loggingLevel.isMoreOrEqualsThan(StreamLoggerLevel.ALL)) {
            Log.i(getTag(tag), message);
        }
        if (loggingHandler != null) {
            loggingHandler.logI(getTag(tag), message);
        }
    }

    public void logD(@NonNull Object tag, @NonNull String message) {
        if (loggingLevel.isMoreOrEqualsThan(StreamLoggerLevel.DEBUG)) {
            Log.d(getTag(tag), message);
        }
        if (loggingHandler != null) {
            loggingHandler.logD(getTag(tag), message);
        }
    }

    public void logW(@NonNull Object tag, @NonNull String message) {
        if (loggingLevel.isMoreOrEqualsThan(StreamLoggerLevel.WARN)) {
            Log.w(getTag(tag), message);
        }
        if (loggingHandler != null) {
            loggingHandler.logW(getTag(tag), message);
        }
    }

    public void logE(@NonNull Object tag, @NonNull String message) {
        if (loggingLevel.isMoreOrEqualsThan(StreamLoggerLevel.ERROR)) {
            Log.e(getTag(tag), message);
        }
        if (loggingHandler != null) {
            loggingHandler.logE(getTag(tag), message);
        }
    }

    public void logT(@NonNull Throwable throwable) {
        if (loggingLevel.isMoreOrEqualsThan(StreamLoggerLevel.ERROR)) {
            throwable.printStackTrace();
        }
        if (loggingHandler != null) {
            loggingHandler.logT(throwable);
        }
    }

    public void logT(@NonNull Object tag, @NonNull Throwable throwable) {
        if (loggingLevel.isMoreOrEqualsThan(StreamLoggerLevel.ERROR)) {
            throwable.printStackTrace();
        }
        if (loggingHandler != null) {
            loggingHandler.logT(getTag(tag), throwable);
        }
    }

    private String getTag(Object tag) {
        if (tag == null) return "null";
        if (tag instanceof String) {
            return (String) tag;
        } else {
            return tag.getClass().getSimpleName();
        }
    }

    public static class Builder {

        private StreamLoggerLevel loggingLevel;
        private StreamLoggerHandler loggingHandler;

        /**
         * Set logging level
         *
         * @param level - Logging {@link StreamLoggerLevel}
         * @return - builder
         */
        public Builder loggingLevel(StreamLoggerLevel level) {
            this.loggingLevel = level;

            return this;
        }

        public Builder setLoggingHandler(StreamLoggerHandler handler) {
            this.loggingHandler = handler;

            return this;
        }

        public StreamChatLogger build() {
            return new StreamChatLogger(loggingLevel, loggingHandler);
        }
    }
}

File: app/src/main/java/io/getstream/chat/android/core/poc/library/logger/StreamChatLogger.kt
package io.getstream.chat.android.core.poc.library.logger

import android.util.Log

class StreamChatLogger constructor(
    private val loggingLevel: StreamLoggerLevel? = StreamLoggerLevel.NOTHING,
    private val loggingHandler: StreamLoggerHandler? = null
) : StreamLogger {

    override fun logT(throwable: Throwable) {
        if (loggingLevel?.isMoreOrEqualsThan(StreamLoggerLevel.ERROR) != false) {
            throwable.printStackTrace()
        }
        loggingHandler?.logT(throwable)
    }

    override fun logT(tag: Any, throwable: Throwable) {
        if (loggingLevel?.isMoreOrEqualsThan(StreamLoggerLevel.ERROR) != false) {
            throwable.printStackTrace()
        }
        loggingHandler?.logT(getTag(tag), throwable)
    }

    override fun logI(tag: Any, message: String) {
        if (loggingLevel?.isMoreOrEqualsThan(StreamLoggerLevel.ALL) != false) {
            Log.i(getTag(tag), message)
        }
        loggingHandler?.logI(getTag(tag), message)
    }

    override fun logD(tag: Any, message: String) {
        if (loggingLevel?.isMoreOrEqualsThan(StreamLoggerLevel.DEBUG) != false) {
            Log.d(getTag(tag), message)
        }
        loggingHandler?.logD(getTag(tag), message)
    }

    override fun logW(tag: Any, message: String) {
        if (loggingLevel?.isMoreOrEqualsThan(StreamLoggerLevel.WARN) != false) {
            Log.w(getTag(tag), message)
        }
        loggingHandler?.logW(getTag(tag), message)
    }

    override fun logE(tag: Any, message: String) {
        if (loggingLevel?.isMoreOrEqualsThan(StreamLoggerLevel.ERROR) != false) {
            Log.e(getTag(tag), message)
        }
        loggingHandler?.logE(getTag(tag), message)
    }

    private fun getTag(tag: Any?): String {
        if (tag == null) return "null"
        return if (tag is String) {
            tag
        } else {
            tag.javaClass.simpleName
        }
    }

    class Builder {
        private var loggingLevel: StreamLoggerLevel? = null
        private var loggingHandler: StreamLoggerHandler? = null

        /**
         * Set logging level
         *
         * @param level - Logging [StreamLoggerLevel]
         * @return - builder
         */

        fun loggingLevel(level: StreamLoggerLevel): Builder {
            loggingLevel = level
            return this
        }

        fun setLoggingHandler(handler: StreamLoggerHandler): Builder {
            loggingHandler = handler
            return this
        }

        fun build(): StreamChatLogger {
            return StreamChatLogger(loggingLevel, loggingHandler)
        }
    }
}